In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
from IPython.display import display
from scipy import stats
from datetime import timedelta
from collections import Counter
import cx_Oracle
#from warnings import filterwarnings
#filterwarnings('ignore')

#### Data loading and processing

In [2]:
data = pd.read_csv('royal_canin_sp_order_positions_pg_q1_and_q2_04092018.dsv', sep=';', low_memory=False)

In [3]:
data.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 147594 entries, 0 to 147593
Data columns (total 34 columns):
SCM_N_ORDER_POS_NR              147594 non-null int64
SCM_N_ORDER_NR                  147594 non-null int64
TDT_T_KEY_ORDER_DATE            147594 non-null object
THR_C_KEY_ORDER_HOUR            147594 non-null object
SIT_N_KEY_SITE                  147594 non-null int64
CUS_N_KEY_CUSTOMER              147594 non-null int64
SCM_SALES_BMF                   147594 non-null object
SCM_SALES_CANCEL_BMF            147594 non-null object
SCM_D_FINANCE_CM1               147594 non-null object
SCM_D_FINANCE_CM2               147594 non-null object
SCM_D_ORIG_FINANCE_ORDER_NET    147594 non-null object
SCM_AP_ID                       147594 non-null int64
SCM_AS_ID                       147594 non-null int64
SCM_MASTER_AP_ID                147594 non-null int64
SCM_IS_ZSP_ORDER                147594 non-null int64
SCM_C_ORDER_STATUS              147594 non-null object
CTP_N_KEY_CUSTOME

In [4]:
data.head()

SCM_N_ORDER_POS_NR  SCM_N_ORDER_NR TDT_T_KEY_ORDER_DATE  \
0           439129985       102960853     21/01/2018 00:00   
1           441361276       103481965     28/01/2018 00:00   
2           483453172       113725396     28/06/2018 00:00   
3           463892437       108829030     16/04/2018 00:00   
4           439950958       103154180     23/01/2018 00:00   

  THR_C_KEY_ORDER_HOUR  SIT_N_KEY_SITE  CUS_N_KEY_CUSTOMER SCM_SALES_BMF  \
0                13:31               7            10289744         24,99   
1                23:39               7             6104684         24,99   
2                16:40               7             9282266          3,49   
3                13:53               7            16515999          7,49   
4                21:34               7             2757473         24,99   

  SCM_SALES_CANCEL_BMF SCM_D_FINANCE_CM1 SCM_D_FINANCE_CM2  \
0                    0            5,7729            2,4986   
1                    0            5,6429            1,8617   
2                    0            1,6143            1,0611   
3                    0            5,1701            4,3297   
4                    0            5,6429            4,5734   

             ...               CZP_EXPIRY_DATE  CZP_ZTQ_ID  CZP_PERCENTAGE  \
0            ...              26/12/2018 23:59         152               8   
1            ...              15/11/2018 23:59         152               8   
2            ...              26/12/2018 23:59         152               8   
3            ...              27/06/2018 23:59         152               8   
4            ...              23/01/2019 23:59         152               8   

   CZP_PRICE    PP_ID   PG_ID  PG4_PARENT_ID PG3_PARENT_ID  \
0        4,9  36702.0  1599.0           47.0           4.0   
1        4,9  36702.0  1599.0           47.0           4.0   
2        4,9  37204.0   914.0           47.0           4.0   
3        4,9  18503.0   215.0           47.0           4.0   
4        4,9  36702.0  1599.0           47.0           4.0   

                                         PP_NAME              AP_SHORT_TEXT  
0            Kaninchentragetasche mit Freigehege  - L 50 x B 54,5 x H 32 cm  
1            Kaninchentragetasche mit Freigehege  - L 50 x B 54,5 x H 32 cm  
2  Trixie ausziehbarer Spieltunnel für Frettchen          - ø 10 × 19/75 cm  
3                        Kletterbaum für Hamster          - 15 x 14 x 14 cm  
4            Kaninchentragetasche mit Freigehege  - L 50 x B 54,5 x H 32 cm  

[5 rows x 34 columns]

In [5]:
def get_date(x):
    return x.split(' ')[0]

In [6]:
data['TDT_T_KEY_ORDER_DATE'] = data['TDT_T_KEY_ORDER_DATE'].map(get_date)

In [7]:
%run -i cleaning_n_formatting.py

Duplicated rows detected:  1060
Start dropping..
Number of order IDs having more than one SP_START_DATE now:  0


In [8]:
data.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 146534 entries, 0 to 147593
Data columns (total 40 columns):
SCM_N_ORDER_POS_NR              146534 non-null int64
SCM_N_ORDER_NR                  146534 non-null int64
TDT_T_KEY_ORDER_DATE            146534 non-null datetime64[ns]
THR_C_KEY_ORDER_HOUR            146534 non-null object
SIT_N_KEY_SITE                  146534 non-null int64
CUS_N_KEY_CUSTOMER              146534 non-null int64
SCM_SALES_BMF                   146534 non-null float64
SCM_SALES_CANCEL_BMF            146534 non-null float64
SCM_D_FINANCE_CM1               146534 non-null float64
SCM_D_FINANCE_CM2               146534 non-null float64
SCM_D_ORIG_FINANCE_ORDER_NET    146534 non-null float64
SCM_AP_ID                       146534 non-null int64
SCM_AS_ID                       146534 non-null int64
SCM_MASTER_AP_ID                146534 non-null int64
SCM_IS_ZSP_ORDER                146534 non-null int64
SCM_C_ORDER_STATUS              146534 non-null object
CTP_

In [9]:
data.head(10)

SCM_N_ORDER_POS_NR  SCM_N_ORDER_NR TDT_T_KEY_ORDER_DATE  \
0           439129985       102960853           2018-01-21   
1           441361276       103481965           2018-01-28   
2           483453172       113725396           2018-06-28   
3           463892437       108829030           2018-04-16   
4           439950958       103154180           2018-01-23   
5           457019742       107178514           2018-03-22   
6           472474517       110988419           2018-05-19   
7           464272219       108924125           2018-04-17   
8           452536377       106125471           2018-03-07   
9           447791830       104997420           2018-02-19   

  THR_C_KEY_ORDER_HOUR  SIT_N_KEY_SITE  CUS_N_KEY_CUSTOMER  SCM_SALES_BMF  \
0                13:31               7            10289744          24.99   
1                23:39               7             6104684          24.99   
2                16:40               7             9282266           3.49   
3                13:53               7            16515999           7.49   
4                21:34               7             2757473          24.99   
5                16:30               7            16455058           7.98   
6                20:15               7             7889866          29.99   
7                18:12               7            10476794          29.99   
8                15:58               7             4221306          29.99   
9                20:53               7            18486370           3.99   

   SCM_SALES_CANCEL_BMF  SCM_D_FINANCE_CM1  SCM_D_FINANCE_CM2    ...      \
0                   0.0             5.7729             2.4986    ...       
1                   0.0             5.6429             1.8617    ...       
2                   0.0             1.6143             1.0611    ...       
3                   0.0             5.1701             4.3297    ...       
4                   0.0             5.6429             4.5734    ...       
5                   0.0             3.0150             2.5416    ...       
6                   0.0            10.2351             8.8613    ...       
7                   0.0            10.1351             8.3117    ...       
8                   0.0            10.4251             9.3381    ...       
9                   0.0             1.5075             1.0757    ...       

   PG4_PARENT_ID  PG3_PARENT_ID  \
0           47.0            4.0   
1           47.0            4.0   
2           47.0            4.0   
3           47.0            4.0   
4           47.0            4.0   
5           47.0            4.0   
6           47.0            4.0   
7           47.0            4.0   
8           47.0            4.0   
9           47.0            4.0   

                                         PP_NAME              AP_SHORT_TEXT  \
0            Kaninchentragetasche mit Freigehege  - L 50 x B 54,5 x H 32 cm   
1            Kaninchentragetasche mit Freigehege  - L 50 x B 54,5 x H 32 cm   
2  Trixie ausziehbarer Spieltunnel für Frettchen          - ø 10 × 19/75 cm   
3                        Kletterbaum für Hamster          - 15 x 14 x 14 cm   
4            Kaninchentragetasche mit Freigehege  - L 50 x B 54,5 x H 32 cm   
5                                   Target Stick                  - 1 Stück   
6            Kaninchentragetasche mit Freigehege  - L 50 x B 54,5 x H 32 cm   
7            Kaninchentragetasche mit Freigehege  - L 50 x B 54,5 x H 32 cm   
8            Kaninchentragetasche mit Freigehege  - L 50 x B 54,5 x H 32 cm   
9                                   Target Stick                  - 1 Stück   

       ORDER_DATETIME WEEK_NUMBER  MONTH_NUMBER FINANCE_ORDER_NET    REL_CM1  \
0 2018-01-21 13:31:00           3             1           25.6694  22.489423   
1 2018-01-28 23:39:00           4             1           26.5951  21.217818   
2 2018-06-28 16:40:00          26             6           93.4712   1.727056   
3 2018-04-16 13:53:00          16             4           3

In [10]:
#Not equal to 0 means the value being cancelled
data['POS_VALID'] = (data['SCM_SALES_CANCEL_BMF'] == 0) * 1 

In [11]:
data[data['POS_VALID'] == 0][['SCM_SALES_BMF', 'SCM_SALES_CANCEL_BMF', 'SCM_D_ORIG_FINANCE_ORDER_NET']].sample(10)

SCM_SALES_BMF  SCM_SALES_CANCEL_BMF  SCM_D_ORIG_FINANCE_ORDER_NET
43790         49.9900               41.3100                       41.3140
92092          3.9800                3.2900                        3.2893
54156          4.4900                3.7100                        3.7107
13844          4.9900                4.1200                        4.1240
8491          19.4900               16.1100                       16.1074
30826         13.9900               11.5600                       11.5620
105226        71.9600               14.8700                       59.4711
82905          7.9900                6.6000                        6.6033
73430          5.2224                4.3178                        4.3160
63032         14.9900               12.3900                       12.3884

In [12]:
data['TAXES'] = data['SCM_SALES_BMF'] / data['SCM_D_ORIG_FINANCE_ORDER_NET']

In [13]:
data[['SCM_SALES_BMF', 'SCM_D_ORIG_FINANCE_ORDER_NET', 'TAXES']].sample(10)

SCM_SALES_BMF  SCM_D_ORIG_FINANCE_ORDER_NET     TAXES
121157          -2.50                       -2.0661  1.210009
124061          -7.40                       -6.1157  1.210000
16237           15.99                       13.2149  1.209998
132252          -5.75                       -4.7521  1.209991
53925           49.99                       41.3140  1.210001
68748           27.99                       23.1322  1.210002
14006           10.49                        8.6694  1.210003
3775             0.00                        0.0000       NaN
57785           96.99                       80.1570  1.210000
108871          -4.40                       -3.6364  1.209988

#### Gross sales, net sales

In [14]:
# Almost 2M € gross sales
np.round((data['SCM_SALES_BMF'] * data['POS_VALID']).sum(), 2)

1927670.29

In [15]:
# 1.6M € net sales
np.round(sum(data['SCM_D_ORIG_FINANCE_ORDER_NET'] * data['POS_VALID']), 2)

1596605.22

#### Revenue of SP sold

In [16]:
data[data['SCM_MASTER_AP_ID'] == 42225]['CZP_PRICE'].value_counts()

0.00    5642
4.90     730
4.99     311
1.90      10
Name: CZP_PRICE, dtype: int64

In [17]:
np.round(data[data['SCM_MASTER_AP_ID'] == 42225]['CZP_PRICE'].sum(), 2)

5147.89

#### SP discount

In [18]:
# 115K €
np.round(data[data['SCM_MASTER_AP_ID'] == 42715]['SCM_SALES_BMF'].sum(), 2)

-114916.81

#### How would it be in case we had brand specific SP where discounts only apply to Brand product groups?

In [19]:
data.columns

Index(['SCM_N_ORDER_POS_NR', 'SCM_N_ORDER_NR', 'TDT_T_KEY_ORDER_DATE',
       'THR_C_KEY_ORDER_HOUR', 'SIT_N_KEY_SITE', 'CUS_N_KEY_CUSTOMER',
       'SCM_SALES_BMF', 'SCM_SALES_CANCEL_BMF', 'SCM_D_FINANCE_CM1',
       'SCM_D_FINANCE_CM2', 'SCM_D_ORIG_FINANCE_ORDER_NET', 'SCM_AP_ID',
       'SCM_AS_ID', 'SCM_MASTER_AP_ID', 'SCM_IS_ZSP_ORDER',
       'SCM_C_ORDER_STATUS', 'CTP_N_KEY_CUSTOMER_TYPE', 'SCM_OCI_LAND',
       'OSP_KO_COUNT_NET', 'OSP_3_LEAD_P_PG_ID', 'OSP_3_LEAD_P_PG_SHARE',
       'OSP_4_LEAD_P_PG_ID', 'OSP_4_LEAD_P_PG_SHARE', 'CZP_START_DATE',
       'CZP_EXPIRY_DATE', 'CZP_ZTQ_ID', 'CZP_PERCENTAGE', 'CZP_PRICE', 'PP_ID',
       'PG_ID', 'PG4_PARENT_ID', 'PG3_PARENT_ID', 'PP_NAME', 'AP_SHORT_TEXT',
       'ORDER_DATETIME', 'WEEK_NUMBER', 'MONTH_NUMBER', 'FINANCE_ORDER_NET',
       'REL_CM1', 'REL_CM2', 'POS_VALID', 'TAXES'],
      dtype='object')

In [20]:
# Example
data[data['SCM_N_ORDER_NR'] == 108350020][['SCM_MASTER_AP_ID', 'SCM_SALES_BMF', 'SCM_D_FINANCE_CM1', 'SCM_D_FINANCE_CM2',
                                          'SCM_D_ORIG_FINANCE_ORDER_NET', 'PG3_PARENT_ID', 'PG4_PARENT_ID',
                                          'OSP_3_LEAD_P_PG_ID', 'OSP_4_LEAD_P_PG_ID', 'PP_NAME', 'AP_SHORT_TEXT']]

SCM_MASTER_AP_ID  SCM_SALES_BMF  SCM_D_FINANCE_CM1  SCM_D_FINANCE_CM2  \
27772              23170         0.0000             0.0000             0.0000   
64427               6855         3.4900             1.5943             1.2422   
68935              54808        49.9900             6.6440             1.7810   
74732              54607         0.0000            -0.3500            -0.6970   
74733              65409         1.5548             0.4218             0.0658   
74734              65409         1.4352             0.3893             0.0607   
88366              79327        28.9900             8.8187             3.9976   
93569              32119         3.9900             1.0375             0.5292   
93570              69053         1.9900             0.4746             0.0662   
101800             62414         2.9900             1.3911             1.0165   
122790              6169         0.0000             0.0000             0.0000   
122791             42715        -4.7200            -3.9008            -3.9008   

        SCM_D_ORIG_FINANCE_ORDER_NET  PG3_PARENT_ID  PG4_PARENT_ID  \
27772                         0.0000         2430.0         9272.0   
64427                         2.8843           34.0         1559.0   
68935                        41.3140          252.0       261925.0   
74732                         0.0000            3.0           34.0   
74733                         1.2850            3.0           34.0   
74734                         1.1861            3.0           34.0   
88366                        23.9587            3.0          301.0   
93569                         3.2975         2068.0         2072.0   
93570                         1.6446         2068.0         2072.0   
101800                        2.4711           36.0         2173.0   
122790                        0.0000            NaN            NaN   
122791                       -3.9008            NaN            NaN   

        OSP_3_LEAD_P_PG_ID  OSP_4_LEAD_P_PG_ID  \
27772                  252              261925   
64427                  252              261925   
68935                  252              261925   
74732                  252              261925   
74733                  252              261925   
74734                  252              261925   
88366                  252              261925   
93569                  252              261925   
93570                  252              261925   
101800                 252              261925   
122790                 252              261925   
122791                 252              261925   

                                 PP_NAME                   AP_SHORT_TEXT  
27772                  Extra Bonuspunkte         - 100 Bonuspunkte extra  
64427                  Trixie Katzenmalz                         - 100 g  
68935          Royal Canin Maine Coon 31    - Maine Coon Adult 10 + 2 kg  
74732         Catessy Knabber-Snack 65 g             Mit Geflügel & Käse  
74733   Smilla Multi-Vitamin Katzenpaste                          - 50 g  
74734            Smilla Malt Katzenpaste                          - 50 g  
88366                PeeWee Wood Pellets                           - 9kg  
93569    Megapack Gourmet Perle 8 x 85 g  - Erlesene Streifen mit Gemüse  
93570    MP/Megapack  - Gourmet 4 x 85 g         - Duetto die Mare Fisch  
101800          Trixie Catnip-Spielspray                        - 175 ml  
122790                               NaN                             NaN  
122791                               NaN                             NaN

+ Reverse the discount applying and get new sales value per order

In [21]:
data[data['SCM_N_ORDER_NR'] == 108350020]['SCM_SALES_BMF'][:11].sum()

94.42999999999999

In [22]:
data[data['SCM_N_ORDER_NR'] == 108350020]['SCM_SALES_BMF'][:12].sum()

89.71

In [23]:
data_without_sp_disc = data[~(data['SCM_MASTER_AP_ID'] == 42715)]

In [24]:
without_sp_disc = data_without_sp_disc.copy()

In [25]:
# Total sales value if there was no SP discount applying on these orders
np.round(sum(without_sp_disc['SCM_SALES_BMF'] * without_sp_disc['POS_VALID']), 2)

2042587.1

+ Locate the suitable products for the specific brand SP

In [26]:
query = """select PPR_N_PHY_PRODUCT_ID
    , PPR_N_PHY_ARTICLE_ID
    , PPR_V_PRODUCT_DESC
    , PPR_V_LOGISTICS_DESC
    , PPR_BRAND
from zoocube.DDSTD_PPR_PHYS_PRODUCT
where PPR_BRAND like '%oyal%'"""

In [27]:
conn = cx_Oracle.connect('zooor', 'zoo12or', 'zoorep', threaded=True, encoding = "UTF-8", nencoding = "UTF-8")
curs = conn.cursor()
pg4ids = pd.read_sql(query, con = conn)
conn.close()

In [28]:
pg4ids

PPR_N_PHY_PRODUCT_ID  PPR_N_PHY_ARTICLE_ID  \
0                     4892                  4528   
1                     4892                  4590   
2                     4923                 22519   
3                     4923                 10110   
4                     4923                 12708   
5                     7625                 22520   
6                     7625                 13049   
7                    18120                 24038   
8                    18120                 24040   
9                    18120                 23695   
10                    7625                  1036   
11                    4911                 22518   
12                    4911                 18349   
13                    4911                 13050   
14                    4911                  5680   
15                    4911                  5666   
16                    4921                 13048   
17                   18120                 24041   
18                   11522                 15319   
19                    5275                  9650   
20                    5275                  5472   
21                    4907                  5109   
22                    4905                  5107   
23                    4905                 10575   
24                    4905                  5104   
25                    4905                  1704   
26                   10812                 12683   
27                   10812                 12686   
28                   19998                 27262   
29                   10334                 13246   
...                    ...                   ...   
2070                 57786                 81169   
2071                 57700                 81167   
2072                 57698                 81093   
2073                 57783                 81166   
2074                 57698                 81165   
2075                 57985                 81533   
2076                 57988                 81536   
2077                 57987                 81535   
2078                 57986                 81534   
2079                 58092                 81537   
2080                 57699                 81544   
2081                 57783                 81543   
2082                 58091                 81441   
2083                 57989                 81442   
2084                 57488                 80749   
2085                 58281                 81799   
2086                 58284                 81802   
2087                 58282                 81800   
2088                 58283                 81801   
2089                 59570                 83013   
2090                 59571                 82974   
2091                 59571                 83016   
2092                 59570                 83012   
2093                 59572                 82976   
2094                 59570                 83014   
2095                 59572                 82975   
2096                 59572                 83017   
2097                 59571                 83015   
2098                 58970                 82592   
2099                 43678                 81692   

                                     PPR_V_PRODUCT_DESC  \
0                       4892 - Royal Canin Medium Puppy   
1                       4892 - Royal Canin Medium Puppy   
2                     4923 - Royal Canin Mini Mature +8   
3                     4923 - Royal Canin Mini Mature +8   
4                     4923 - Royal Canin Mini Mature +8   
5                7625 - Royal Canin Mini Digestive Care   
6                7625 - Royal Canin Mini Digestive Care   
7                       18120 - Welpenset (ohne Futter)   
8                       18120 - Welpenset (ohne Futter)   
9                       18120 - Welpenset (ohne Futter)   
10               7625 - Royal Canin Mini Digestive Care   
11                        4911 - Royal Canin Mini Puppy   
12            

In [29]:
rc_prods = list(set(pg4ids['PPR_N_PHY_ARTICLE_ID']))

In [30]:
rc_names = list(without_sp_disc[without_sp_disc['SCM_AP_ID'].isin(rc_prods)]['PP_NAME'].unique())

In [31]:
len(rc_names)

185

In [32]:
rc_names

['Royal Canin Nutritional Supplement',
 'Royal Canin - Vet Diet Hypoallergenic',
 'Royal Canin Veterinary Diet Dog - Sensitive Control',
 'Royal Canin Vet Diet - Urinary S/O',
 'RC Veterinary Diet Gastro Intestinal Junior',
 'Royal Canin Vet Diet - Satiety Support weight management',
 'Royal Canin Veterinary Diet Dog - Gastro Intestinal',
 'Royal Canin Veterinary Diet Dog - Mobility C2P+',
 'Royal Canin Veterinary Diet Anallergenic',
 'RC Veterinary Diet Skin Care Small',
 'RC Veterinary Urinary S/O MC',
 'Royal Canin Medium Light Weight Care',
 'RC Veterinary Diet Hypoallergenic Small',
 'RC Veterinary Diet Hypoallergenic MC',
 'Royal Canin Veterinary Diet Dog - Hepatic',
 'RC Veterinary Diet Gastro Intestinal Low Fat',
 'Royal Canin Veterinary Diet Dog - Obesity Mangement',
 'Royal Canin Veterinary Diet Fibre Response',
 'RC Veterinary Diet Skin Care',
 'Royal Canin Veterinary Diet Dental Special Small',
 'RC Veterinary Diet Satiety Small',
 'Royal Canin Veterinary Diet Dog - Dental'

In [33]:
# List every product group from Royal Canin...
#rc_prod_groups = list(without_sp_disc[without_sp_disc['PP_NAME'].fillna('-').str.contains('Royal Canin')]['PP_NAME'].unique()) +\
#                 list(without_sp_disc[without_sp_disc['PP_NAME'].fillna('-').str.contains('RC')]['PP_NAME'].unique())

In [34]:
#len(rc_prod_groups)

In [35]:
#test_included_rc_prods = [i for i in rc_prod_groups if 'Vet' not in i ]

In [36]:
#len(test_included_rc_prods)

In [37]:
# Let's exclude from the list all products belonging to Veterinary Diets
included_rc_prods = [i for i in rc_names  if 'Vet' not in i ]

In [38]:
len(included_rc_prods)

133

In [39]:
without_sp_disc['DISC_POS'] = (without_sp_disc['PP_NAME'].isin(included_rc_prods)) * 1

In [40]:
without_sp_disc[['PP_NAME', 'SCM_SALES_BMF', 'DISC_POS', 'CZP_PERCENTAGE']].sample(20)

PP_NAME  SCM_SALES_BMF  \
55335                       Royal Canin Kitten Sterilised        42.9900   
100497                                 Advantage CAT (NL)        32.9900   
1264                   Catit Design Senses Massage-Center         0.0000   
18680                               Biokat's Classic 3in1         7.9933   
70345                    Royal Canin British Shorthair 34        27.9900   
93015                     Megapack Gourmet Perle 8 x 85 g         0.0000   
13484            Royal Canin - Vet Diet - Satiety Support        64.9900   
118999                                                NaN        -0.3500   
82714   Catit Water Fountain Trinkbrunnen 3 l inkl. Fu...        30.9933   
29473                    Savic Bag it Up Litter Tray Bags         5.9800   
121675                                                NaN         0.0000   
83440                  Animonda vom Feinsten Kitten 100 g         3.6900   
53127                           Royal Canin Sterilised 37        51.9900   
36537      Royal Canin Veterinary Diet Megapack 12 x 85 g        11.9900   
52135                              Royal Canin Outdoor 30        83.9700   
30247                    Savic Bag it Up Litter Tray Bags         4.9900   
28003                                   Extra Bonuspunkte         0.0000   
66725              ProDen PlaqueOff Zahnpflege für Katzen        12.9900   
77157                    Smilla Multi-Vitamin Katzenpaste         0.0000   
5183                            Gra-Mix Ardenner Mischung        16.9900   

        DISC_POS  CZP_PERCENTAGE  
55335          1               5  
100497         0               5  
1264           0               8  
18680          0               5  
70345          1               5  
93015          0               5  
13484          0               5  
118999         0               5  
82714          0               5  
29473          0               8  
121675         0               5  
83440          0               5  
53127          1               5  
36537          0               5  
52135          1               5  
30247          0               5  
28003          0               5  
66725          0               5  
77157          0               5  
5183           0               5

+ Apply SP discount only on the suitable order positions

In [41]:
without_sp_disc['NEW_SALES_VALUE'] = without_sp_disc['SCM_SALES_BMF'] - \
((without_sp_disc['SCM_SALES_BMF'] * without_sp_disc['CZP_PERCENTAGE'] / 100) * without_sp_disc['DISC_POS'])

In [42]:
without_sp_disc[['PP_NAME', 'SCM_SALES_BMF', 'DISC_POS', 'CZP_PERCENTAGE', 'NEW_SALES_VALUE']].sample(20)

PP_NAME  SCM_SALES_BMF  \
103360                          Royal Canin Pouches, 85 g         34.990   
24913                                        Savings Plan          0.000   
33340                               Sheba Megapack 12x85g          4.990   
28794                           toothbrush holder zooplus          0.000   
19100                     Biokats Micro Fresh Katzenstreu         14.990   
2324                          LED Pointer Catch the Light          2.490   
63850                                Fellpflege Handschuh          0.000   
111381                                                NaN          0.990   
99023                                    Felix Snacks 60g          0.000   
40843                                    Litter Locker II          6.990   
55090                             Royal Canin Sensible 33         49.990   
17475                               Biokat's Classic 3in1          7.194   
29066                Royal Canin Breed Pug/ Mops 25 Adult         41.990   
56648                              Royal Canin Outdoor 30         26.990   
50456                               Royal Canin Indoor 27         24.990   
123147                                                NaN          0.000   
30958                               Cabrio Katzentoilette         17.990   
30518                              Trixie Trocken-Shampoo          3.990   
17183                           BIO Yarrah Cat Food Fisch         49.990   
13943   Feringa Trockenfutter Ente, Kürbis,Cranberries...         14.990   

        DISC_POS  CZP_PERCENTAGE  NEW_SALES_VALUE  
103360         1               5          33.2405  
24913          0               5           0.0000  
33340          0               5           4.9900  
28794          0               5           0.0000  
19100          0               5          14.9900  
2324           0               5           2.4900  
63850          0               5           0.0000  
111381         0               8           0.9900  
99023          0               5           0.0000  
40843          0               5           6.9900  
55090          1               5          47.4905  
17475          0               8           7.1940  
29066          1               5          39.8905  
56648          1               5          25.6405  
50456          1               5          23.7405  
123147         0               5           0.0000  
30958          0               5          17.9900  
30518          0               5           3.9900  
17183          0               5          49.9900  
13943          0               5          14.9900

In [43]:
without_sp_disc['BASKET_VALUE'] = (without_sp_disc['SCM_SALES_BMF'] * without_sp_disc['POS_VALID']).groupby(without_sp_disc['SCM_N_ORDER_NR']).transform('sum')

In [44]:
without_sp_disc['O_POS_SHARE'] = (without_sp_disc['SCM_SALES_BMF'] * without_sp_disc['POS_VALID']) / without_sp_disc['BASKET_VALUE']

In [45]:
data[data['SCM_N_ORDER_NR'] == 101550829]

SCM_N_ORDER_POS_NR  SCM_N_ORDER_NR TDT_T_KEY_ORDER_DATE  \
53988            433251159       101550829           2018-01-01   
132801           433251160       101550829           2018-01-01   

       THR_C_KEY_ORDER_HOUR  SIT_N_KEY_SITE  CUS_N_KEY_CUSTOMER  \
53988                 02:49               7            10197386   
132801                02:49               7            10197386   

        SCM_SALES_BMF  SCM_SALES_CANCEL_BMF  SCM_D_FINANCE_CM1  \
53988           49.99                   0.0             7.9940   
132801          -2.50                   0.0            -2.0661   

        SCM_D_FINANCE_CM2    ...                     PP_NAME  AP_SHORT_TEXT  \
53988              2.9740    ...     Royal Canin Sensible 33        - 10 kg   
132801            -2.0661    ...                         NaN            NaN   

            ORDER_DATETIME  WEEK_NUMBER  MONTH_NUMBER FINANCE_ORDER_NET  \
53988  2018-01-01 02:49:00            1             1           39.2479   
132801 2018-01-01 02:49:00            1             1           39.2479   

          REL_CM1   REL_CM2  POS_VALID     TAXES  
53988   20.367969  7.577475          1  1.210001  
132801  -5.264231 -5.264231          1  1.210009  

[2 rows x 42 columns]

#### With new brand specific scenario...

##### Gross sales, net sales

In [46]:
# Almost 2M € gross sales
np.round(sum(without_sp_disc['NEW_SALES_VALUE'] * without_sp_disc['POS_VALID']), 2)

1981221.14

In [47]:
# 1.6M € net sales
np.round((without_sp_disc['NEW_SALES_VALUE'] / without_sp_disc['TAXES'] * without_sp_disc['POS_VALID']).sum(), 2)

1641065.68

##### Revenue of SP sold

In [48]:
without_sp_disc[without_sp_disc['SCM_MASTER_AP_ID'] == 42225]['CZP_PRICE'].value_counts()

0.00    5642
4.90     730
4.99     311
1.90      10
Name: CZP_PRICE, dtype: int64

In [49]:
np.round(without_sp_disc[without_sp_disc['SCM_MASTER_AP_ID'] == 42225]['CZP_PRICE'].sum(), 2)

5147.89

##### SP discount

In [50]:
# 62K €
np.round((without_sp_disc['SCM_SALES_BMF'] - without_sp_disc['NEW_SALES_VALUE']).sum(), 2)

61566.7

|  | Gross sales | Net sales | Revenue of SP sold | SP discount |
| --- | --- | --- | --- |
| OLD RC SP | 1927670 € | 1596605 € | 5148 € | 114917 € | 
| Brand-specific RC SP | 1981221 € | 1641066 € | 5148 € | 61567 € |
|Delta| 2.78% | 2.78% | 0% | -46.42% |

Deltas have been calculated like this: (new state - previous state) / previous state * 100

In [51]:
# Orders including Brand Specific Products 
without_sp_disc[(without_sp_disc['DISC_POS'] == 1) & (without_sp_disc['POS_VALID'] == 1)]['SCM_N_ORDER_NR'].nunique()

22212

In [52]:
# All orders
without_sp_disc['SCM_N_ORDER_NR'].nunique()

33072

In [53]:
# % from Orders having Brand specific products
np.round(without_sp_disc[(without_sp_disc['DISC_POS'] == 1) & (without_sp_disc['POS_VALID'] == 1)]['SCM_N_ORDER_NR'].nunique() /\
without_sp_disc['SCM_N_ORDER_NR'].nunique() * 100, 2)

67.16